modules/xobjdetect/src/main.cpp

```diff
@@ -0,0 +1,556 @@
+    }
+sampling_finished:
+    return patches;
+}
+
+void train_boosting()
+{
+    cerr << "read imgs" << endl;
+    vector<Mat> pos_imgs = read_imgs("/home/vlad/gsoc/lfw_faces");
+    assert(pos_imgs.size());
+
+    const char neg_path[] = "/home/vlad/gsoc/rtsd_bg";
+    vector<Mat> neg_imgs = sample_patches(neg_path, 24, 24, pos_imgs.size());
+    assert(neg_imgs.size());
+
+    int n_features;
+    Mat pos_data, neg_data;
+
+    Ptr<CvFeatureEvaluator> eval = CvFeatureEvaluator::create(CvFeatureParams::LBP);
+    eval->init(CvFeatureParams::create(CvFeatureParams::LBP), 1, Size(24, 24));
+    n_features = eval->getNumFeatures();
+
+    const int stages[] = {32, 64, 128, 256, 512, 1024, 2048, 4096};
+    const int stage_count = sizeof(stages) / sizeof(*stages);
+    const int stage_neg = 5000;
+    const int max_per_image = 25;
+
+    const float scales_arr[] = {.3, .4, .5, .6, .7, .8, .9, 1};
+    const vector<float> scales(scales_arr,
+            scales_arr + sizeof(scales_arr) / sizeof(*scales_arr));
+
+    WaldBoost boost;
+    vector<String> neg_filenames;
+    glob(neg_path, neg_filenames);
+
+
+    for (int i = 0; i < stage_count; ++i) {
+
+        cerr << "compute features" << endl;
+
+        pos_data = Mat1b(n_features, pos_imgs.size());
+        neg_data = Mat1b(n_features, neg_imgs.size());
+
+        for (size_t k = 0; k < pos_imgs.size(); ++k) {
+            eval->setImage(pos_imgs[k], +1, 0, boost.get_feature_indices());
+            for (int j = 0; j < n_features; ++j) {
+                pos_data.at<uchar>(j, k) = (*eval)(j, 0);
+            }
+        }
+
+        for (size_t k = 0; k < neg_imgs.size(); ++k) {
+            eval->setImage(neg_imgs[k], 0, 0, boost.get_feature_indices());
+            for (int j = 0; j < n_features; ++j) {
+                neg_data.at<uchar>(j, k) = (*eval)(j, 0);
+            }
+        }
+
+
+        boost.reset(stages[i]);
+        boost.fit(pos_data, neg_data);
+
+        if (i + 1 == stage_count) {
+            break;
+        }
+
+        int bootstrap_count = 0;
+        int img_i = 0;
+        for (; img_i < neg_filenames.size(); ++img_i) {
+            cerr << "win " << bootstrap_count << "/" << stage_neg
+                 << " img " << (img_i + 1) << "/" << neg_filenames.size() << "\r";
+            Mat img = imread(neg_filenames[img_i], CV_LOAD_IMAGE_GRAYSCALE);
+            vector<Rect> bboxes;
+            Mat1f confidences;
+            boost.detect(eval, img, scales, bboxes, confidences);
+
+            if (confidences.rows > 0) {
+                Mat1i indices;
+                sortIdx(confidences, indices,
+                        CV_SORT_EVERY_COLUMN + CV_SORT_DESCENDING);
+
+                int win_count = min(max_per_image, confidences.rows);
+                win_count = min(win_count, stage_neg - bootstrap_count);
+                Mat window;
+                for (int k = 0; k < win_count; ++k) {
+                    resize(img(bboxes[indices(k, 0)]), window, Size(24, 24));
```

Comment:
bboxes.size() may not equal to confidences.rows, because bboxes are grouped in function boost.detect, this leads to an access exception in line 509, e.g. bboxes.size() is 0 and confidences.rows is 1